src/concurrency/lock_manager.cpp

```diff
@@ -0,0 +1,254 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// lock_manager.cpp
+//
+// Identification: src/concurrency/lock_manager.cpp
+//
+// Copyright (c) 2015-18, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <map>
+
+#include <boost/thread/shared_mutex.hpp>
+#include "concurrency/lock_manager.h"
+
+namespace peloton {
+namespace concurrency {
+
+//===--------------------------------------------------------------------===//
+// Lock manager
+//===--------------------------------------------------------------------===//
+
+/**
+ * @brief    Get the global variable instance of lock manager
+ * @return   Pointer to the global lock manager instance
+ */
+LockManager *LockManager::GetInstance() {
+  static LockManager global_lm;
+  return &global_lm;
+}
+
+/**
+ * @brief    Initialize lock for given oid
+ * @details  Initialize lock for given oid. Currently,
+ * only supports read/write lock
+ * @param    oid  The oid of data table
+ * @param    type The type of lock to initialize
+ * @return   bool true if successful, false if not
+ */
+bool LockManager::InitLock(oid_t oid, LockManager::LockType /*type*/) {
+  // Need to lock internal lock to protect internal lock map
+  internal_rw_lock_.lock();
+
+  // Try to find the lock in table
+  boost::upgrade_mutex *lock = GetLock(oid);
+  if (lock != nullptr) {
+    internal_rw_lock_.unlock();
+    LOG_TRACE("Init lock failed.");
+    return false;
+  }
+
+  // Initialize new lock for the object
+  boost::upgrade_mutex *rw_lock = new boost::upgrade_mutex();
+
+  std::pair<oid_t, boost::upgrade_mutex *> p;
+  p.first = oid;
+  p.second = rw_lock;
+  // Insert lock into map
+  lock_map_.insert(p);
+
+  // Unlock internal lock
+  internal_rw_lock_.unlock();
+  LOG_TRACE("Init lock success.");
+  return true;
+}
+
+/**
+ * @brief    Remove the lock specified by oid from data structure
+ * @param    oid  The oid of data table
+ * @return   bool true if successful, false if not
+ */
+bool LockManager::RemoveLock(oid_t oid) {
+  // Need to lock internal lock to protect internal lock map
+  internal_rw_lock_.lock();
+
+  // Try to access the lock
+  boost::upgrade_mutex *rw_lock = GetLock(oid);
+  if (rw_lock == nullptr) {
+    internal_rw_lock_.unlock();
+    LOG_TRACE("Remove lock failed.");
+    return false;
+  }
+
+  // Remove from the lock map
+  lock_map_.erase(oid);
+  delete (rw_lock);
```

Comment:
Do we need to check whether there are still transactions holding the lock pointer here?
I think using a deprecated way to delete a lock is better as when removing a lock there might still be transacations holding the lock pointer. Maybe add a TODO here and we can help handle this.

Reply:
Good idea. Maybe make it a shared_ptr, as it will automatically clean itself after the reference count decreased to 0. Thanks for the suggestion!

Reply:
About the automatic lock release.. Yes I know the method and it can ensure the lock be released if some error occurs in the middle of function execution. But it's indeed a little tricky and not standard c++ code style, so I don't know if it's applicable.. Maybe we need to ask Andy about this?